org.eclipse.m2e.tests/src/org/eclipse/m2e/tests/jdt/JavaClasspathTest.java

```diff
@@ -328,4 +332,66 @@ public void test480137_addingResourceFoldersAreNotRemovedDuringConfigurationUpda
     }, //
         javaProject.getRawClasspath());
   }
+
+  public void test525880_JavaAccessRulesArePreservedOnUpdate() throws Exception {
+    IProject project = importProject("projects/525880_java_pom_with_access_rules/pom.xml");
+    addAccessRulesToClasspathContainers(JavaCore.create(project));
+    assertContainersHaveAccessRules(JavaCore.create(project));
+
+    updateProjectConfiguration(project);
+
+    assertContainersHaveAccessRules(JavaCore.create(project));
+  }
+
+  private void updateProjectConfiguration(IProject project) throws CoreException, InterruptedException {
+    MavenPlugin.getProjectConfigurationManager().updateProjectConfiguration(project, monitor);
+    waitForJobsToComplete();
+  }
+
+  private void assertContainersHaveAccessRules(IJavaProject javaProject) throws JavaModelException {
+    List<IClasspathEntry> containerEntries = Arrays.asList(javaProject.getRawClasspath()).stream()
```

Comment:
Stream.of(javaProject.getRawClasspath()) is shorter :-)

Reply:
You're right!  Are you asking me to change this?  If so, are there other changes you're looking for too, or is this it?

Reply:
I haven't tested the changes yet, but looking at the test, it looks like you're testing the maven container retains the access rules, globally. 
I think it'd better to be more explicit, test access rules on a particular dependency. Eg. add junit and something else, add access rules to junit, check only junit has access rules after update.

Reply:
I think I'm missing something.  

This change is about ensuring that access rules added to the M2E container are preserved.  E.g. 

```

	<classpathentry kind="con" path="org.eclipse.m2e.MAVEN2_CLASSPATH_CONTAINER">
		<attributes>
			<attribute name="maven.pomderived" value="true"/>
		</attributes>
		<accessrules>
			<accessrule kind="discouraged" pattern="**/internal/**"/>
		</accessrules>
	</classpathentry>
```

So this test is testing only the functionality that was added.

Reply:
ok my bad, as I said I haven't checked yet. But wouldn't it be better to also have a more fine grained control over each dependency?

Reply:
Perhaps, but that is not our use-case.

Out of interest I've tried adding per-dependency access rules to dependencies in the M2E container in the Eclipse UI.  The UI appears to work, but no changes appear in the .classpath file, and when I re-open the "configure build path" dialog, the access rules are gone.  So I'm wondering if such a thing is even possible?

Reply:
Yes it's possible this would require more complex changes to serialize the state of the maven classpath container (those changes would not be stored in .classpath, but in another file under .metadata). Let's put that one on the back burner.

Reply:
OK, sounds good.